WIP: Dynamic Providers blog post (#2174)
pulumi/docs: content/blog/dynamic-providers/index.md

---
title: "Dynamic Providers"
date: 2019-12-20

> meta_desc: "Pulumi dynamic providers manage resources that do not have prebuilt providers such as AWS, Azure, and GCP"

Reviewer: ```suggestion
meta_desc: "Pulumi dynamic providers manage resources that are not covered in pre-built providers"
```

meta_image: meta.png
authors:
    - praneet-loke
tags:
    - providers
    - resources
---

Pulumi has several resource providers (over 45!) that allow you to interact with your favorite cloud provider or resource using a programming language. But what if there isn’t a provider available for what you want to accomplish? That’s where dynamic providers can help you. They can be an extremely important part of your infrastructure app.

<!--more-->

> A provider manages the CRUD lifecycle of a resource against some underlying service. The underlying service can be one of the public clouds or an on-premise service. For example, AWS, Azure, GCP, and Kubernetes are all resource providers, which can create/update/destroy resources in the respective clouds.

Reviewer 2: Consider adding a section with a distinct title to explain what providers are, maybe add a picture of how Pulumi works

Reviewer 3: Hmm I do have a section below this that is exactly meant for a brief intro to providers. But I didn't go into too much details on purpose. I could link to [this](https://www.pulumi.com/docs/intro/concepts/how-pulumi-works/) page and maybe even include the diagram from there.

> ## _Dynamic_ Resource Provider

Reviewer 2: Nit: italic looks odd to me here

Reviewer 3: It is to emphasize the difference in the title between this section and the previous one.

Reviewer 3: Now that enough time has passed, I do agree with you that this looks odd.

> > **Note**: Dynamic providers are currently only supported in Javascript, Typescript, and Python.

Reviewer: ```suggestion
> **Note**: Dynamic providers are currently only supported in JavaScript, TypeScript, and Python.
```

> Pulumi provides appropriate lifecycle callbacks through the `pulumi.dynamic.ResourceProvider` abstract class. While Pulumi does not know what sort of resources a dynamic provider will create, the lifecycle hooks must return data in specific formats, which Pulumi uses as hints to call the appropriate lifecycle hooks.

Reviewer: ```suggestion
Pulumi provides appropriate life-cycle callbacks through the `pulumi.dynamic.ResourceProvider` abstract class. While Pulumi does not know what sort of resources a dynamic provider will create, the life-cycle hooks must return data in specific formats, which Pulumi uses as data to call the appropriate callbacks. 
```

> For example, if the user changes an input property to your dynamic resource, the `diff` hook will be called with both the old (if applicable) and new inputs. You can then compare to see which input properties have changed and give hints to Pulumi as to whether `delete` should be called or whether to call the `update` hook.

Reviewer 2: Would love to see a picture with all hooks listed, which would give the intuition of the component lifecycle.

Reviewer 3: I agree a picture with the hooks would be a idea. Let me see if I can come up with something.

Reviewer 3: I added the same diagram from our How Pulumi Works page. Do you think we should still include another diagram here?

> The `check` lifecycle hook will be called before any other hook to let you validate the inputs and populate defaults, if necessary.

When a dynamic resource instance updates, the `delete` hook will be called to allow you to perform any necessary cleanups before the dynamic resource is removed from your stack’s state.

> ### Provider Serialization

Reviewer 2: IMO, this is too short to be easily understandable, and maybe not very relevant at this point in the article. Maybe add a section of how dynamic providers work internally closer to the end of the article?

Reviewer 3: Hmm it is a sub-section. Maybe I can remove the title and just make it a regular paragraph under the parent section?

Reviewer 3: I removed this section because it does feel too short.

It is important to note that once an instance of the dynamic resource is created, Pulumi serializes the provider and stores it in your [state](https://www.pulumi.com/docs/intro/concepts/state/#state) file. When you change the source code of your dynamic resource, the serialized version is updated.

## Use Cases

> Dynamic providers can be used for almost any task. Here are some use cases:

Reviewer 2: I would discuss these possibilities as we go through examples, and then generalized in the conclusion at the end of the article

Run a command-line tool when a dependent resource is created.
Copy files to a VM once the VM is created.

> Call REST APIs in each lifecycle hook (CRUD) to adopt an external resource under the cloud resource model.

Reviewer: ```suggestion
Call REST APIs in each life-cycle hook (CRUD) to adopt an external resource under the cloud resource model.
```

> Call REST APIs in each lifecycle hook (CRUD) to adopt an external resource under the cloud resource model.

Reviewer 2: Consider making a bullet-point list

Reviewer 3: Done

> ### Examples

Reviewer 2: I would try to structure all examples in similar ways:

- The task at hand
- Why dynamic provider is needed for this task
- Snippet
- Links to full code

That should help to understand the scenario.

Reviewer 3: Great idea!

Reviewer 3: I have done this.

#### Enable Azure Storage's Static Websites Feature

[Here's](https://github.com/pulumi/examples/blob/master/azure-ts-static-website/staticWebsite.ts) the source for this example. The following is an excerpt from the [example](https://github.com/pulumi/examples/blob/master/azure-ts-static-website/index.ts) where the dynamic provider is actually used. Note that the Azure Storage Account is created before using this dynamic provider.

```ts
const staticWebsite = new StorageStaticWebsite("website-static", {

> 	accountName: storageAccount.name,

Reviewer 2: I would add a storage account declaration to the snippet

Reviewer 3: Done

});
```

#### Add A Custom Domain To An Azure CDN endpoint

Along with adding a custom domain to the CDN endpoint, this dynamic provider also enables HTTPS provided by Azure's one-click [HTTPS enablement](https://docs.microsoft.com/en-us/azure/cdn/cdn-custom-ssl?tabs=option-1-default-enable-https-with-a-cdn-managed-certificate).

> As before, some of the details, such as the creation of the CDN profile and its endpoint are omitted here for brevity. You can check out the full example [here](https://github.com/pulumi/examples/tree/master/azure-ts-dynamicresource).

Reviewer: ```suggestion
As before, some of the details, such as the creation of the CDN profile and its endpoint are omitted here for clarity. You can check out the full example [here](https://github.com/pulumi/examples/tree/master/azure-ts-dynamicresource).
```
